### ISSUE TITLE
[Service Bus] Batch Delete updates for GA

### ISSUE THREAD

**Issue by jsquire** [Azure/azure-sdk-for-net#43801]
# Summary

The preview version of the Service Bus batch delete feature has a number of known service bugs which the client must work around.  Before the feature is released to a stable version, the client should remove the work-arounds and validate the service fixes.

## Work arounds

- `ServiceBusReceiver.DeleteMessagesAsync` is currently validating the maximum message count allowed by the service.  As this is an arbitrary limit that may change, the client should not restrict the upper bound for the `messageCount` parameter and allow the service to validate.

- The implementation of `ServiceBusReceiver.PurgeMessagesAsync` accounts for the service not meeting its guarantee that the requested `messageCount` will be deleted if there are enough messages eligible for deletion, instead deleting an inconsistent amount.  Once this is fixed, the implementation should be adjusted as shown in [Receiver updates](#receiver-updates) and the unit tests from "ReceiverTests" should be updated as shown in [ReceiverTests updates](#receivertests-updates).

### Receiver updates

```csharp
        /// <summary>
        /// Attempts to purge all messages from an entity.  Locked messages are not eligible for removal and
        /// will remain in the entity.
        /// </summary>
        /// <param name="beforeEnqueueTimeUtc">An optional <see cref="DateTimeOffset"/> representing the cutoff time for deletion. Only messages that were enqueued before this time will be deleted.  If not specified, <see cref="DateTimeOffset.UtcNow"/> will be assumed.</param>
        /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
        /// <remarks>
        /// If the lock for a message is held by a receiver, it will be respected and the message will not be deleted.
        ///
        /// This method may invoke multiple service requests to delete all messages.  As a result, it may exceed the configured <see cref="ServiceBusRetryOptions.TryTimeout"/>.
        /// If you need control over the amount of time the operation takes, it is recommended that you pass a <paramref name="cancellationToken"/> with the desired timeout set for cancellation.
        ///
        /// Because multiple service requests may be made, the possibility of partial success exists.  In this scenario, the method will stop attempting to delete additional messages
        /// and throw the exception that was encountered.  It is recommended to evaluate this exception and determine which messages may not have been deleted.
        /// </remarks>
        /// <returns>The number of messages that were deleted.</returns>
        public virtual async Task<int> PurgeMessagesAsync(
            DateTimeOffset? beforeEnqueueTimeUtc = null,
            CancellationToken cancellationToken = default)
        {
            beforeEnqueueTimeUtc ??= DateTimeOffset.UtcNow;

            var deleted = await DeleteMessagesAsync(MaxDeleteMessageCount, beforeEnqueueTimeUtc.Value, cancellationToken).ConfigureAwait(false);

            if (deleted == MaxDeleteMessageCount)
            {
               var batchCount = MaxDeleteMessageCount;

               while (batchCount == MaxDeleteMessageCount)
               {
                   batchCount = await DeleteMessagesAsync(MaxDeleteMessageCount, beforeEnqueueTimeUtc.Value, cancellationToken).ConfigureAwait(false);
                   deleted += batchCount;
               }
            }

            return deleted;
        }

        /// <summary>
        /// Deletes up to <paramref name="messageCount"/> messages from the entity. The actual number
        /// of deleted messages may be less if there are fewer eligible messages in the entity.
        /// </summary>
        /// <param name="messageCount">The desired number of messages to delete.  This value is limited by the service and governed <see href="https://learn.microsoft.com/azure/service-bus-messaging/service-bus-quotas">Service Bus quotas</see>.</param>
        /// <param name="beforeEnqueueTimeUtc">An optional <see cref="DateTimeOffset"/> representing the cutoff time for deletion. Only messages that were enqueued before this time will be deleted.  If not specified, <see cref="DateTimeOffset.UtcNow"/> will be assumed.</param>
        /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
        /// <returns>The number of messages that were deleted.</returns>
        /// <remarks>If the lock for a message is held by a receiver, it will be respected and the message will not be deleted.</remarks>
        /// <exception cref="ArgumentOutOfRangeException">
        /// Occurs when the <paramref name="messageCount"/> is less than 1 or exceeds the maximum allowed, as determined by the Service Bus service.
        /// For more information on service limits, see <see href="https://learn.microsoft.com/azure/service-bus-messaging/service-bus-quotas#messaging-quotas"/>.
        /// </exception>
        public virtual async Task<int> DeleteMessagesAsync(
            int messageCount,
            DateTimeOffset? beforeEnqueueTimeUtc = null,
            CancellationToken cancellationToken = default)
        {
            Argument.AssertAtLeast(messageCount, 1, nameof(messageCount));
            Argument.AssertNotDisposed(IsDisposed, nameof(ServiceBusReceiver));
            _connection.ThrowIfClosed();

            beforeEnqueueTimeUtc ??= DateTimeOffset.UtcNow;
            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();

            Logger.DeleteMessagesStart(Identifier, messageCount, beforeEnqueueTimeUtc.Value);

            using DiagnosticScope scope = ClientDiagnostics.CreateScope(
                DiagnosticProperty.DeleteActivityName,
                ActivityKind.Client,
                MessagingDiagnosticOperation.Settle);

            scope.Start();

            int numMessagesDeleted;
            try
            {
                numMessagesDeleted = await InnerReceiver.DeleteMessagesAsync(messageCount, beforeEnqueueTimeUtc.Value, cancellationToken).ConfigureAwait(false);
            }
            catch (Exception exception)
            {
                Logger.DeleteMessagesException(Identifier, exception.ToString());
                scope.Failed(exception);
                throw;
            }

            Logger.DeleteMessagesComplete(Identifier, numMessagesDeleted);
            return numMessagesDeleted;
        }
```

### ReceiverTests updates

```csharp
        [Test]
        public async Task PurgeMessagesHandlesLowCount()
        {
            var expectedDeleteCount = 30;
            var mockReceiver = new Mock<ServiceBusReceiver>() { CallBase = true };
            var receiver = mockReceiver.Object;

            mockReceiver
                .Setup(receiver => receiver.DeleteMessagesAsync(
                    It.IsAny<int>(),
                    It.IsAny<DateTimeOffset>(),
                    It.IsAny<CancellationToken>()))
                .ReturnsAsync(expectedDeleteCount);

            // Delete with no parameters should continue to invoke the service
            // operation until the count of messages deleted is less than the
            // maximum allowed.

            var deleteCount = await receiver.PurgeMessagesAsync();
            Assert.AreEqual(expectedDeleteCount, deleteCount);

            mockReceiver
                .Verify(receiver => receiver.DeleteMessagesAsync(
                    ServiceBusReceiver.MaxDeleteMessageCount,
                    It.IsAny<DateTimeOffset>(),
                    It.IsAny<CancellationToken>()),
                    Times.Once);
        }

        [Test]
        public async Task PurgeMessagesEvaluatesReturnedCount()
        {
            var expectedDeleteCount = (ServiceBusReceiver.MaxDeleteMessageCount * 3) - 1;
            var mockReceiver = new Mock<ServiceBusReceiver>() { CallBase = true };
            var receiver = mockReceiver.Object;

            mockReceiver
                .SetupSequence(receiver => receiver.DeleteMessagesAsync(
                    It.IsAny<int>(),
                    It.IsAny<DateTimeOffset>(),
                    It.IsAny<CancellationToken>()))
                .ReturnsAsync(ServiceBusReceiver.MaxDeleteMessageCount)
                .ReturnsAsync(ServiceBusReceiver.MaxDeleteMessageCount)
                .ReturnsAsync(ServiceBusReceiver.MaxDeleteMessageCount - 1);

            // Delete with no parameters should continue to invoke the service
            // operation until the count of messages deleted is less than the
            // maximum allowed.

            var deleteCount = await receiver.PurgeMessagesAsync();
            Assert.AreEqual(expectedDeleteCount, deleteCount);

            mockReceiver
                .Verify(receiver => receiver.DeleteMessagesAsync(
                    ServiceBusReceiver.MaxDeleteMessageCount,
                    It.IsAny<DateTimeOffset>(),
                    It.IsAny<CancellationToken>()),
                    Times.Exactly(3));
        }

        [Test]
        public async Task PurgeMessagesForDateHandlesLowCount()
        {
            var expectedDeleteCount = 2;
            var expectedDate = new DateTimeOffset(2015, 10, 27, 0, 0, 0, 0, TimeSpan.Zero);
            var mockReceiver = new Mock<ServiceBusReceiver>() { CallBase = true };
            var receiver = mockReceiver.Object;

            mockReceiver
                .Setup(receiver => receiver.DeleteMessagesAsync(
                    It.IsAny<int>(),
                    It.IsAny<DateTimeOffset>(),
                    It.IsAny<CancellationToken>()))
                .ReturnsAsync(expectedDeleteCount);

            // Purge for a date should continue to invoke the service
            // operation until the count of messages deleted is less than the
            // maximum allowed.

            var deleteCount = await receiver.PurgeMessagesAsync(expectedDate);
            Assert.AreEqual(expectedDeleteCount, deleteCount);

            mockReceiver
                .Verify(receiver => receiver.DeleteMessagesAsync(
                    ServiceBusReceiver.MaxDeleteMessageCount,
                    expectedDate,
                    It.IsAny<CancellationToken>()),
                    Times.Once);
        }

        [Test]
        public async Task PurgeMessagesForDateEvaluatesReturnedCount()
        {
            var expectedDeleteCount = (ServiceBusReceiver.MaxDeleteMessageCount * 4) -1;
            var expectedDate = new DateTimeOffset(2015, 10, 27, 0, 0, 0, 0, TimeSpan.Zero);
            var mockReceiver = new Mock<ServiceBusReceiver>() { CallBase = true };
            var receiver = mockReceiver.Object;

            mockReceiver
                .SetupSequence(receiver => receiver.DeleteMessagesAsync(
                    It.IsAny<int>(),
                    It.IsAny<DateTimeOffset>(),
                    It.IsAny<CancellationToken>()))
                .ReturnsAsync(ServiceBusReceiver.MaxDeleteMessageCount)
                .ReturnsAsync(ServiceBusReceiver.MaxDeleteMessageCount)
                .ReturnsAsync(ServiceBusReceiver.MaxDeleteMessageCount)
                .ReturnsAsync(ServiceBusReceiver.MaxDeleteMessageCount - 1);

            // Delete for a date should continue to invoke the service
            // operation until the count of messages deleted is less than the
            // maximum allowed.

            var deleteCount = await receiver.PurgeMessagesAsync(expectedDate);
            Assert.AreEqual(expectedDeleteCount, deleteCount);

            mockReceiver
                .Verify(receiver => receiver.DeleteMessagesAsync(
                    ServiceBusReceiver.MaxDeleteMessageCount,
                    expectedDate,
                    It.IsAny<CancellationToken>()),
                    Times.Exactly(4));
        }

        [Test]
        public async Task DeleteMessagesForCountPassesTheCurrentDate()
        {
            var expectedCount = 2000;
            var mockConnection = ServiceBusTestUtilities.CreateMockConnection();
            var mockTransportReceiver = new Mock<TransportReceiver>();

            mockConnection
                .Setup(connection => connection.CreateTransportReceiver(
                    It.IsAny<string>(),
                    It.IsAny<ServiceBusRetryPolicy>(),
                    It.IsAny<ServiceBusReceiveMode>(),
                    It.IsAny<uint>(),
                    It.IsAny<string>(),
                    It.IsAny<string>(),
                    It.IsAny<bool>(),
                    It.IsAny<bool>(),
                    It.IsAny<CancellationToken>()))
                .Returns(mockTransportReceiver.Object);

            mockTransportReceiver
                .Setup(receiver => receiver.DeleteMessagesAsync(
                    It.IsAny<int>(),
                    It.IsAny<DateTimeOffset>(),
                    It.IsAny<CancellationToken>()))
                .ReturnsAsync(expectedCount);

            // Delete for a count should use the date/time so that
            // the operation cannot loop forever. Capture a call time just
            // prior to calling the method so that we can infer the value
            // used by the method is just slightly in the future.

            var callTime = DateTimeOffset.UtcNow;
            var receiver = new ServiceBusReceiver(mockConnection.Object, "fake", false, new ServiceBusReceiverOptions());
            var returnedCount = await receiver.DeleteMessagesAsync(expectedCount);

            Assert.AreEqual(expectedCount, returnedCount);

            mockTransportReceiver
                .Verify(receiver => receiver.DeleteMessagesAsync(
                    expectedCount,
                    It.Is<DateTimeOffset>(value => value >= callTime),
                    It.IsAny<CancellationToken>()),
                    Times.Once);
        }

        [Test]
        public async Task DeleteMessagesPassesParametersToTransport()
        {
            using var cancellationSource = new CancellationTokenSource();

            var requestedCount = 2500;
            var expectedCount = 2000;
            var expectedDate = new DateTimeOffset(2015, 10, 27, 0, 0, 0, 0, TimeSpan.Zero);
            var mockConnection = ServiceBusTestUtilities.CreateMockConnection();
            var mockTransportReceiver = new Mock<TransportReceiver>();

            mockConnection
                .Setup(connection => connection.CreateTransportReceiver(
                    It.IsAny<string>(),
                    It.IsAny<ServiceBusRetryPolicy>(),
                    It.IsAny<ServiceBusReceiveMode>(),
                    It.IsAny<uint>(),
                    It.IsAny<string>(),
                    It.IsAny<string>(),
                    It.IsAny<bool>(),
                    It.IsAny<bool>(),
                    It.IsAny<CancellationToken>()))
                .Returns(mockTransportReceiver.Object);

            mockTransportReceiver
                .Setup(receiver => receiver.DeleteMessagesAsync(
                    It.IsAny<int>(),
                    It.IsAny<DateTimeOffset>(),
                    It.IsAny<CancellationToken>()))
                .ReturnsAsync(expectedCount);

            var receiver = new ServiceBusReceiver(mockConnection.Object, "fake", false, new ServiceBusReceiverOptions());
            var returnedCount = await receiver.DeleteMessagesAsync(requestedCount, expectedDate, cancellationSource.Token);
            Assert.AreEqual(expectedCount, returnedCount);

            mockTransportReceiver
                .Verify(receiver => receiver.DeleteMessagesAsync(
                    requestedCount,
                    expectedDate,
                    cancellationSource.Token),
                    Times.Once);
        }
```

**Comment by 0xced** [comment]
The `PurgeMessagesAsync` method is public in version 7.18.0-beta.1 but is internal in the latest version (7.20.1 as of writing).

Was this an overlook? Can we use the `PurgeMessagesAsync` method like [suggested](https://github.com/Azure/azure-service-bus/issues/1#issuecomment-2610632089) by @ramonsmits (using reflection/UnsafeAccessor) or are there still issues waiting to be solved before we can safely use it?

**Comment by IanKemp** [comment]
> The `PurgeMessagesAsync` method is public in version 7.18.0-beta.1 but is internal in the latest version (7.20.1 as of writing).
> 
> Was this an overlook? Can we use the `PurgeMessagesAsync` method like [suggested](https://github.com/Azure/azure-service-bus/issues/1#issuecomment-2610632089) by [@ramonsmits](https://github.com/ramonsmits) (using reflection/UnsafeAccessor) or are there still issues waiting to be solved before we can safely use it?

This is bizarre to me:

* The latest `ServiceBusReceiver` code in this repo shows these methods as `public` and was last changed 2+ years ago
* The docs on learn.microsoft.com show these methods as `public` as of version 7.18.0-beta.1, which according to nuget.org was released 2024-05-08
* Yet the current GA version 7.20.1, released 2025-06-12 i.e. which is (or should be) **newer** than 7.18.0-beta.1, has both methods as `internal`
* Trying to figure out exactly what version of the code **actually** shipped in what GA version is impossible, because you guys apparently stopped tagging Microsoft.Azure.ServiceBus releases at version 5.2.0

What is the point of having source code publicly available if it doesn't reflect the reality of what's released? What is the point of having source code publicly available if it's impossible to determine what was released and when?

In short, this is an insane mess and I'd really appreciate if you'd fix it _yesterday_.

**Comment by IanKemp** [comment]
Also, `PurgeMessagesAsync` should not return `Task<int>` but `Task<long>`.
